e2e/clients/dashboard-tus/app.js

```diff
@@ -10,7 +10,7 @@ import '@uppy/dashboard/dist/style.css'
 const companionUrl = 'http://localhost:3020'
 const uppy = new Uppy()
   .use(Dashboard, { target: '#app', inline: true })
-  .use(Tus, { endpoint: 'https://tusd.tusdemo.net/files' })
+  .use(Tus, { endpoint: 'https://tusd.tusdemo.net/files', retryStatusCodes: [429, 401] })
```

Comment:
I get that it's probably just a random value you picked to test that feature, but it makes no sense to do exponential backoff when receiving a 401 response code. 425 would be a much better value imo, we can also use 418 to make clear that we are just testing.

Reply:
The idea is that your auth token is expired and you use onBeforeRequest to generate a new token. Makes sense?

Reply:
I'm not sure it's the correct fix for this problem, there's no reason to do exponential backoff in this use case. What if instead we expose a `onShouldRetry` option?

Reply:
Good point, I figured we don't want to get rid of our `onShouldRetry` as you don't want to write exponential backoff yourself, but I suppose if you want something custom that's the cost.

Reply:
I think we can do both, by passing the default handler as second argument (but of course that would be problematic if Tus ever introduces a second argument of their own, ugh)

Reply:
We could do something similar to `onBeforeRequest`:

```js
        if (typeof opts.onBeforeRequest === 'function') {
          opts.onBeforeRequest(req)
        }
```

but instead do it like so: 

```js
        if (typeof opts.onShouldRetry === 'function') {
          const shouldRetry = opts.onShouldRetry(req)
          if (shouldRetry) {
            return shouldRetry
          }
        }
```

This would make it sort of consistent with how we do `onBeforeRequest` and if your custom logic doesn't want to retry, it will still go into our retrying logic for 429.

What do you think?

Reply:
I'm suggesting a fifth solution:

```js
if (opts.onShouldRetry != null) { // if there's a user-supplied `onShouldRetry` option:
  uploadOptions.onShouldRetry = (err) => opts.onShouldRetry(err, defaultOnShouldRetry);
} else {
 uploadOptions.onShouldRetry = defaultOnShouldRetry;
}
```

In the case of the request in the linked issue, the implementation would look something like:

```js
.use(Tus, {
  endpoint: 'https://tusd.tusdemo.net/files',
  onShouldRetry(err, next) {
    if (err?.originalResponse?.getStatus() === 401) return true;
    return next(err);
})
```

It would be different than `onBeforeRequest`, because users who supply this option should be in control on whether they want the default behavior (they can decide not to rate limit when getting a 429), while in `onBeforeRequest` we must be in control otherwise it can overflow the `limit` option (that user can set to `Infinity` to opt-out).

Regarding `2.`, not sure what that would look like, do you have a suggestion?